impl/spanner/directory/directory.go

```diff
@@ -0,0 +1,221 @@
+
+var directoriesCols = []string{"DirectoryID", "Map", "Log", "VRFPublicKey", "VRFPrivateKey", "MinInterval", "MaxInterval", "Deleted"}
+
+var _ directory.Storage = &Table{}
+
+// Table gives access to the directory table.
+type Table struct {
+	db *ktspanner.Database
+}
+
+// New returns a directory.Storage client backed by Spanner.
+func New(db *ktspanner.Database) *Table {
+	return &Table{db: db}
+}
+
+// spannerDirectory represents one row in the Directories table in spanner.
+type spannerDirectory struct {
+	DirectoryID   string
+	Map           []byte
+	Log           []byte
+	VRFPublicKey  []byte
+	VRFPrivateKey []byte
+	MinInterval   int64
+	MaxInterval   int64
+	Deleted       bool
+}
+
+func readRow(row *spanner.Row) (*directory.Directory, error) {
+	var r spannerDirectory
+	if err := row.ToStruct(&r); err != nil {
+		return nil, err
+	}
+
+	vrfPriv, err := unwrapAnyProto(r.VRFPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var tmap tpb.Tree
+	if err := proto.Unmarshal(r.Map, &tmap); err != nil {
+		return nil, err
+	}
+	var tlog tpb.Tree
+	if err := proto.Unmarshal(r.Log, &tlog); err != nil {
+		return nil, err
+	}
+
+	return &directory.Directory{
+		DirectoryID: r.DirectoryID,
+		Map:         &tmap,
+		Log:         &tlog,
+		VRF:         &keyspb.PublicKey{Der: r.VRFPublicKey},
+		VRFPriv:     vrfPriv,
+		MinInterval: time.Duration(r.MinInterval) * time.Nanosecond,
+		MaxInterval: time.Duration(r.MaxInterval) * time.Nanosecond,
+		Deleted:     r.Deleted,
+	}, nil
+}
+
+// List returns all Directories. showDeleted indicates whether deleted directories should be returned.
+func (t *Table) List(ctx context.Context, showDeleted bool) ([]*directory.Directory, error) {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+
+	ret := []*directory.Directory{}
+	err = rtx.Read(ctx, table, spanner.AllKeys(), directoriesCols).Do(
+		func(r *spanner.Row) error {
+			d, err := readRow(r)
+			if err != nil {
+				return err
+			}
+			if d.Deleted && !showDeleted {
+				return nil
+			}
+			ret = append(ret, d)
+			return nil
+		})
+	return ret, err
+}
+
+// Write creates a new Directory.
+func (t *Table) Write(ctx context.Context, dir *directory.Directory) error {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return err
+	}
+	// Prepare data.
+	anyPB, err := ptypes.MarshalAny(dir.VRFPriv)
+	if err != nil {
+		return err
+	}
+	anyData, err := proto.Marshal(anyPB)
+	if err != nil {
+		return err
+	}
+
+	tmap, err := proto.Marshal(dir.Map)
+	if err != nil {
+		return err
+	}
+	tlog, err := proto.Marshal(dir.Log)
+	if err != nil {
+		return err
+	}
+
+	m, err := spanner.InsertStruct(table, spannerDirectory{
+		DirectoryID:   dir.DirectoryID,
+		Map:           tmap,
+		Log:           tlog,
+		VRFPublicKey:  dir.VRF.GetDer(),
+		VRFPrivateKey: anyData,
+		MinInterval:   dir.MinInterval.Nanoseconds(),
+		MaxInterval:   dir.MaxInterval.Nanoseconds(),
+		Deleted:       false,
```

Comment:
Not feeling strongly, but this field could be just omitted?

Reply:
Yep